lib/core/src/Cardano/Wallet/StakePool/Metrics.hs

```
@@ -14,37 +16,169 @@ module Cardano.Wallet.StakePool.Metrics
+    , combineMetrics
+
+    , StakePoolLayer (..)
+    , newStakePoolLayer
+    , ErrListStakePools (..)
     )
     where
 
 import Prelude
 
+import Cardano.BM.Trace
+    ( Trace, appendName, logInfo )
+import Cardano.Wallet.Network
+    ( NetworkLayer (..), follow )
 import Cardano.Wallet.Primitive.Types
-    ( BlockHeader (..), PoolId (..), SlotId (..) )
+    ( BlockHeader (..), EpochNo, PoolId (..), SlotId (..), Stake )
+import Control.Concurrent
+    ( forkIO )
+import Control.Concurrent.MVar
+    ( MVar, modifyMVar_, newMVar, readMVar )
+import Control.Monad
+    ( void )
+import Control.Monad.IO.Class
+    ( liftIO )
+import Control.Monad.Trans.Except
+    ( ExceptT (..), throwE, withExceptT )
+import Data.List.NonEmpty
+    ( NonEmpty )
+import Data.Map.Merge.Strict
+    ( WhenMissing, mergeA, traverseMissing, zipWithMatched )
 import Data.Map.Strict
     ( Map )
-import Data.Word
-    ( Word64 )
+import Data.Text
+    ( Text )
 import Fmt
     ( Buildable (..), blockListF', fmt, (+|), (|+) )
 import GHC.Generics
     ( Generic )
 
 import qualified Data.Map.Strict as Map
 
+--
+-- Metrics
+--
+
+data ErrListStakePools
+    = ErrListStakePoolsStakeIsUnreachable
+    | ErrListStakePoolsMetricsIsUnsynced
+    | ErrListStakePoolsInternalError ErrListStakePoolsInternalErr
+
+data ErrListStakePoolsInternalErr
+    = ErrListStakePoolInternalErrInconsistentData
+    | ErrListStakePoolInternalErrReadStateErr
+
+newtype StakePoolLayer m = StakePoolLayer
+    { listStakePools
+        :: ExceptT ErrListStakePools m [(PoolId, (Stake, Int))]
+    }
+
+newStakePoolLayer
+    :: NetworkLayer IO tx (BlockHeader, PoolId)
+    -> Trace IO Text
+    -> IO (StakePoolLayer IO)
+newStakePoolLayer nl tr = do
+    mvar <- worker nl tr
+    return $ StakePoolLayer
+        { listStakePools = Map.toList <$>
+            combineMetrics
+                (withE1 $ stakeDistribution nl)
+                (liftIO $ readMVar mvar)
+        }
+  where
+    withE1 = withExceptT $ const ()
```

So we are discarding an error here :thinking: ?